src/super_gradients/training/models/detection_models/customizable_detector.py

```diff
@@ -0,0 +1,199 @@
+        c3_out_channels, c4_out_channels, c5_out_channels = in_channels
+
+        NeckStage1Cls = arch_params.neck1.type
+        NeckStage2Cls = arch_params.neck2.type
+        NeckStage3Cls = arch_params.neck3.type
+        NeckStage4Cls = arch_params.neck4.type
+
+        self.neck1 = NeckStage1Cls(arch_params.neck1, [c5_out_channels, c4_out_channels])
+        self.neck2 = NeckStage2Cls(arch_params.neck2, [self.neck1.out_channels[1], c3_out_channels])
+        self.neck3 = NeckStage3Cls(arch_params.neck3, [self.neck2.out_channels[1], self.neck2.out_channels[0]])
+        self.neck4 = NeckStage4Cls(arch_params.neck4, [self.neck3.out_channels, self.neck1.out_channels[0]])
+
+        self.out_channels = [
+            self.neck2.out_channels[1],
+            self.neck3.out_channels,
+            self.neck4.out_channels,
+        ]
+
+    def forward(self, inputs):
+        c3, c4, c5 = inputs
+
+        x_n1_inter, x = self.neck1([c5, c4])
+        x_n2_inter, p3 = self.neck2([x, c3])
+        p4 = self.neck3([p3, x_n2_inter])
+        p5 = self.neck4([p4, x_n1_inter])
+
+        return p3, p4, p5
+
+
+class ThreeHeads(nn.Module):
+    """
+    Apply three heads and combine predictions into the shape expected by SG detection losses
+    """
+    def __init__(self, arch_params: Union[HpmStruct, DictConfig], in_channels: List[int]):
+        super().__init__()
+        arch_params = self._pass_num_classes(arch_params)
+
+        Head1Cls = arch_params.head1.type
+        Head2Cls = arch_params.head2.type
+        Head3Cls = arch_params.head3.type
+
+        self.head1 = Head1Cls(arch_params.head1, in_channels[0])
+        self.head2 = Head2Cls(arch_params.head2, in_channels[1])
+        self.head3 = Head3Cls(arch_params.head3, in_channels[2])
+
+    @staticmethod
+    def _pass_num_classes(arch_params: HpmStruct):
+        for i in range(3):
+            arch_params[f'head{i + 1}'].num_classes = arch_params.num_classes
+        return arch_params
+
+    def forward(self, inputs):
+        p3, p4, p5 = inputs
+        return self.combine_preds([self.head1(p3), self.head2(p4), self.head3(p5)])
+
+    def combine_preds(self, preds):
+        outputs = []
+        outputs_logits = []
+        for output, output_logits in preds:
+            outputs.append(output)
+            outputs_logits.append(output_logits)
+
+        return outputs if self.training else (torch.cat(outputs, 1), outputs_logits)
+
+
+class CustomizableDetector(SgModule):
+    """
+    A customizable detector with backbone -> neck -> heads
+
+    A type of each submodule must be defined explicitly. The defaults are:
+      * backbone - FourStageBackbone
+      * neck - PANNeck
+      * heads - ThreeHeads
+
+    By default, initializes BatchNorm eps to 1e-3, momentum to 0.03 and sets all activations to be inplace
+    """
+
+    def __init__(self, arch_params: Union[HpmStruct, DictConfig], type_mapping: Dict = None):
+        """
+        :param arch_params:
```

Comment:
Can you please include an example of arch_params configuration file to instantiate, say, YoloX architecture using this class?
This would help new users to quickly grasp the concept and make their learning experience better.

Reply:
YoloX based on this class isn't ready yet: submodules with the required signature and neck stages modules don't exist in the codebase. It can be done later after this base proves worthy, because it will require weights surgery on the existing pre-trained weights. 
For now I'm refactoring SSD to be based on this class, and can also share internal code that uses this base in private